src/Kestrel.Core/Internal/Http/Http1Connection.cs

```diff
@@ -413,67 +413,77 @@ protected override string CreateRequestId()
         protected override MessageBody CreateMessageBody()
             => Http1MessageBody.For(_httpVersion, HttpRequestHeaders, this);
 
-        protected override async Task<bool> ParseRequestAsync()
+        protected override void InitializeParsing()
```

Comment:
I also think that calling InitializeParsing at the beginning of the first call to BeginRead would be nice for simplifying the HttpProtocol class. Just be sure to only call it when _requestProcessingStatus is RequestPending.

Reply:
If that's a reliable way to detect the first call to BeginRead, sure.

Reply:
It's reliable. There are four ways for ReadAsync to complete.

1. With request data: If the ReadableBuffer is anything but empty, it will move out of RequestPending and into the ParsingRequestLine state.
2. With a graceful connection close: This of course exits the while loop by returning true from the TryParseRequest method and setting the endConnection out param.
3. With a connection error: ReadAsync throws and exits the loop.
4. Because CancelPendingRead: This is the most unusual scenario. CancelPendingRead is only called on the connection's Input pipe during server shutdown. In this scenario, _keepAlive will be set to false first. The loop will exit due to the `else if (!_keepAlive && _requestProcessingStatus == RequestProcessingStatus.RequestPending)` condition in TryParseRequest.

I'm only going into such excruciating detail b/c you're ramping up on Kestrel.

Reply:
Looks like it's actually `Reset` that sets the state back to `RequestPending`, so that wouldn't quite work as-written. Instead, I've moved the code to reset the state up to the `ProcessRequestsAsync` loop itself.

Reply:
Though to be fair `Reset` is on `HttpProtocol`. So in theory we could run it in the outer loop rather than in `BeginRead`. Need to see where HTTP/2 does it to see how that compares. The timeout stuff is HTTP/1.1 only (for now)